### ISSUE TITLE
DOTNET 6 Error -   Cannot find compilation library location for package 'System.Security.Cryptography.Pkcs'

### ISSUE THREAD

**Issue by jacodv** [toddams/RazorLight#460]
**Describe the bug**
Existing tests fail after upgrade.

**To Reproduce**
Upgrade Test Project from Dotnet 5 to Dotnet 6:

**Expected behavior**
The template should transform

**Information (please complete the following information):**
 - OS: Windows 11
 - Dotnet6
 - RazorLight version: 
 -- https://www.nuget.org/packages/RazorLight/2.0.0-rc.3
 -- https://www.nuget.org/packages/RazorLight.NetCore3/
 - 
 - Are you using the OFFICIAL RazorLight package? https://www.nuget.org/packages/razorlight
 - Visual Studio version : Visual Studio Community 2022 Professional Released

**Additional context**
TEST
```[Test]
    public void Convert_GivenValidInput_ShouldConvert()
    {
      // arrange
      Setup();
      var data = new { Name = "TheName" };
      var template = "Name is @Model.Name";

      // action
      var result = _engine.Convert(data, template, true);

      // assert
      result .Should().Be("Name is TheName");
  }
```

IMPLEMETATION:
```
        var result = engine.CompileRenderStringAsync(GetTemplateCachedId(templateData), templateData, data,  (ExpandoObject)null).Result;

```
  
  ERROR:
```

IIAB.Core.Common.Exceptions.RazorParsingException : One or more errors occurred. (Cannot find compilation library location for package 'System.Security.Cryptography.Pkcs')
   at IIAB.Razor.RazorTemplateEngine.Convert(Object data, String templateData, Boolean ifErrorTrySerializeAndDeserialize) in C:\Git\IIAB-netcore\src\IIAB.Razor\RazorTemplateEngine.cs:line 102
   at IIAB.Razor.Tests.RazorTemplateEngineTests.Convert_GivenValidInput_ShouldConvert() in C:\Git\IIAB-netcore\src\IIAB.Razor.Tests\RazorTemplateEngineTests.cs:line 32

```

**Comment by jzabroski** [comment]
@jacodv I don't think RazorLight.Tests has any test with Convert_GivenValidInput_ShouldConvert.  Can you reproduce this in the current repositories tests?

**Comment by jacodv** [comment]
I have cloned the repo, but Visual Studio 2022 keep on hanging trying to load the projects.  I will create a clean solution and try load them again.  If I succeed then I will push to GitHub and revert ASAP.

**Comment by jacodv** [comment]
@jzabroski I have created a repo (https://github.com/jacodv/RazorLightDotNet6)

I get different errors, but they can be fixed by updating the RazorLight.csproj.  Here are my tests in order (branches): 
* branch_jdevil-Dotnet5-Success (Started with a .Net5 Test Project)
* branch_jdevil-Dotnet6-UpdateProjectOnly (Only change the test project to .Net6) - See README in root for error
* branch_jdevil-Dotnet6-ChangeLanguageVersion-To-9-Sucess (Leave Test Project on .Net6, Change LanguageVersion to 9)
* branch_jdevil-Donet6-Add-6-To-RazorLight-Project-success (Update RazorLight project with .Net 6 conditions)

**Comment by jacodv** [comment]
@jzabroski : My Error has todo with a dependency (System.Security.Cryptography.Pkcs) of referenced Nuget packages.  It would seem that different Nuget packages depend of different versions of this dependency.  I realized it by looking at the project.assets.json file in the obj folder.

Three questions:
* Is there a way to for the load of a specific version when the RazorEngine compiles?
* Why does this only occur when the test project (Entry Assembly) is .Net6?
* Is there a way to debug and determine which Nuget package causes the error?  Bear in mind that the razor template and data does not reference any of the nuget packages, it must happen in the load of the reference assembly.

**Comment by adimoldovan92** [comment]
I am having the same problem after upgrading from .NET Core 3.1 to .NET 6. I haven't managed yet to find a fix for this,

**Comment by jacodv** [comment]
@jzabroski I managed to replicate the problem.  All I had todo is add a package reference to `System.Security.Cryptography.Pkcs` in the test project. 

https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Add-IIAB-To-See-Why-Error-Occurs

Thank you in advance

**Comment by jzabroski** [comment]
@jacodv ...Why are you adding a reference to that project?

That said:

> It would seem that different Nuget packages depend of different versions of this dependency. I realized it by looking at the project.assets.json file in the obj folder.

The only solution to fix this conflict is to write an assembly resolve helper.  This isn't really a RazorLight problem, this is a your-dependencies-are-in-hell-you-need-to-sort-it-out problem.

**Comment by jzabroski** [comment]
> I get different errors, but they can be fixed by updating the RazorLight.csproj. Here are my tests in order (branches):
> 
> * branch_jdevil-Dotnet5-Success (Started with a .Net5 Test Project)
> * branch_jdevil-Dotnet6-UpdateProjectOnly (Only change the test project to .Net6) - See README in root for error
> * branch_jdevil-Dotnet6-ChangeLanguageVersion-To-9-Sucess (Leave Test Project on .Net6, Change LanguageVersion to 9)
> * branch_jdevil-Donet6-Add-6-To-RazorLight-Project-success (Update RazorLight project with .Net 6 conditions)

Thanks, I'll try to take a look at this soon to understand the problem.

**Comment by jacodv** [comment]
> @jacodv ...Why are you adding a reference to that project?

@jzabroski I added the reference to show that it is not DLL hell as suggested.  If I only add that project and add RazorLight then I get the error.  Let us assume I had this code in my project

Commit to: https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Add-IIAB-To-See-Why-Error-Occurs

```
[Test]
		public void Convert_GivenValidInput_ShouldConvert()
		{
			// arrange
			Setup();
			var data = new { Name = "TheName" };
			var template = "Name is @Model.Name";

			// action
			var result = _engine.Convert(data, template, true);

			var signedResult = _sign(Encoding.UTF8.GetBytes(result), X509Certificate2.CreateFromPemFile("personal.pem"));

			// assert
			result.Should().Be("Name is TheName");
		}
		
private static byte[] _sign(byte[] message, X509Certificate2 cert)
		{
			var contentInfo = new ContentInfo(message);
			var signedCms = new SignedCms(contentInfo);
			var cmsSigner = new CmsSigner(cert)
			{
				IncludeOption = X509IncludeOption.EndCertOnly
			};

			signedCms.ComputeSignature(cmsSigner);

			var signed = signedCms.Encode();
			return signed;
		}
```

**Comment by jacodv** [comment]
@jzabroski / @toddams,  did you manage to see what might cause this issue?  On my last commit I showed that my initial guess that a 3rd party is causing the problem was incorrect.

**Comment by jzabroski** [comment]
Sorry man.  I have not given this repo a lot of love the last ~2 years (basically, since covid hit).  I acknowledge your repro and think it is great you went that far.  I'm hesitant to sink a lot of time into issues related to loading third party DLLs, since ultimately RazorLight is doing very simple things.  

I would guess there is an issue with the Metadata Resolver, for which there is already an FAQ entry.  

Typically when I debug these problems I try to see if:

1. There is a conflict across versions due to "diamong dependencies" across nuget packages, and there is no way to "roll forward".
2. The problem is due to packaging issues.  Sometimes, OOB (Microsoft term for packages that ship on nuget.org instead of with a target framework SDK that sits on disk) packages don't have a clear package under certain conditions, like netstandard packages vs a specific, concrete target framework (e.g., net5.0).
3. I check to see how things are being packaged.  As I've explained to people before, if you want to use this project on newer frameworks, self-contained deployment is the way to go, but you do open yourself up to security risks due to the underlying dependencies being unpatched nuget.org patches (which is likely part of the reason why Microsoft moved to making the framework something that sits on disk vs. can be included off nuget.org).

**Comment by jzabroski** [comment]
I think, the fix might be to change the way we packaged netstandard2.0, as it is probably ambiguous which version of Microsoft.AspNetCore.* assemblies we should use when we request a frameworkreference for net6.0 in a context where the interstitial project between the entrypoint assembly and the RazorLight assembly is a netstandard2.0 assembly/csproj.

e.g.

```
Program.csproj [net6.0]
 \
  References EmailRenderer.csproj
|
EmailRenderer.csproj [netstandard2.0]
 \
  Package Reference RazorLight
  \
   FrameworkReference 6.0
 \
  Microsoft.AspNetCore.* references 2.1 version
```

In this case, MSBuild will resolve half of the dependencies to netstandard2.0 direct depenencies,& will resolve framework references to net6.0 dependencies provided by Microsoft on Disk for whichever version of the 6.0 SDK you have.

THIS IS JUST A WORKING THEORY I HAVE NOT FULLY TESTED.  Just sharing on my lunch break to try to help unblock you.

**Comment by jacodv** [comment]
@jzabroski , Firstly I appreciate you and the rest of the team's efforts to maintain a crucial function for us.  If I can help, I will do so greatly.

Your assessment seems on track,  It has todo with conflicts  between 2.0 and 2.1.  We need 2.1 for some of the collectable assembly functionality and there lie the issue I believe.  Because .Net 5 works without any issues.

I will also play with your suggestions later today when I have a gap in my day

**Comment by jzabroski** [comment]
The primary reason we support netstandard2.0 TFM is really for .NET Framework 4.8 projects. The Microsoft.AspNetCore.* assemblies do not target net4.8 TFM, so that is why we targeted netstandard2.0.

One possible solution is to create a RazorLight.Net48 project and move the netstandard2.0 dependency there.  It would share the same code as RazorLight.csproj through auto-globbing rules, but have a different assembly title and package name.

**Comment by jzabroski** [comment]
Here are the packages we use and a quick review of which ones keep up to date on versions:

# .NETCoreApp 3.1
| Package | RazorLight Version | Status | Available OOB? | Notes
| - | - | - | - | - |
| [Microsoft.AspNetCore.Mvc.Razor.Extensions](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Razor.Extensions/#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.CodeAnalysis.Razor](https://www.nuget.org/packages/Microsoft.CodeAnalysis.Razor/#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.Extensions.Caching.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Abstractions/#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.Extensions.Caching.Memory](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Memory/#versions-tab) | (>= 3.1.5) |  6.0.0 | ✔️ | - |
| [Microsoft.Extensions.DependencyInjection](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection/#versions-tab) |  (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.Extensions.DependencyModel](https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.Extensions.FileProviders.Physical](https://www.nuget.org/packages/Microsoft.Extensions.FileProviders.Physical#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [Microsoft.Extensions.Primitives](https://www.nuget.org/packages/Microsoft.Extensions.Primitives#versions-tab) | (>= 3.1.5) | 6.0.0 | ✔️ | - |
| [System.Buffers](https://www.nuget.org/packages/System.Buffers#versions-tab) | (>= 4.5.0) | 4.5.1 | ✔️ ❓  | It looks like this was moved to the System.Memory.dll in .NET 6 |

I didn't do the other TFMs, but you get the idea of my approach.  I am not sure if System.Buffers 4.5.1 is the right version for .NET 6.0.  Can you please research that?

**Comment by jzabroski** [comment]
There's also this advisory in the .NET 6 Breaking Changes guide: https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/older-framework-versions-dropped

> Starting with .NET 6 Preview 5, the core libraries packages can no longer be installed into projects whose target framework is older than:
> 
> * .NET Framework 4.6.1
> * .NET Core 3.1
> * .NET Standard 2.0

**Comment by jzabroski** [comment]
In particular, see this:

https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/older-framework-versions-dropped#affected-apis

> # Affected APIs
> The following packages no longer ship old frameworks:
> 
> * **Microsoft.Extensions.DependencyModel**
> * Microsoft.Win32.Registry.AccessControl
> * Microsoft.Win32.SystemEvents
> * System.ComponentModel.Annotations
> * System.ComponentModel.Composition
> * System.ComponentModel.Composition.Registration
> * System.Composition.AttributedModel
> * System.Composition.Convention
> * System.Composition.Hosting
> * System.Composition.Runtime
> * System.Composition.TypedParts
> * System.Data.Odbc
> * System.Diagnostics.DiagnosticSource
> * System.Diagnostics.EventLog
> * System.Diagnostics.PerformanceCounter
> * System.DirectoryServices
> * System.DirectoryServices.AccountManagement
> * System.DirectoryServices.Protocols
> * System.Drawing.Common
> * System.IO.Packaging
> * System.IO.Pipelines
> * System.Management
> * System.Net.Http.WinHttpHandler
> * System.Reflection.Context
> * System.Runtime.Caching
> * System.Runtime.CompilerServices.Unsafe
> * System.Security.Cryptography.Cng
> * System.Security.Cryptography.OpenSsl
> * System.Security.Cryptography.Pkcs
> * System.Security.Cryptography.ProtectedData
> * System.ServiceProcess.ServiceController
> * System.Speech
> * System.Text.Encoding.CodePages
> * **System.Text.Encodings.Web**
> * System.Threading.AccessControl
> * System.Threading.Channels
> 
> The following packages will no longer be updated because their implementation is now part of the .NET 6 platform:
> 
> * Microsoft.Win32.Registry
> * System.ComponentModel.Annotations
> * **System.IO.FileSystem.AccessControl**
> * System.IO.Pipes.AccessControl
> * System.Security.AccessControl
> * System.Security.Cryptography.Cng
> * System.Security.Cryptography.OpenSsl
> * System.Security.Principal.Windows

I bolded three packages that I think are interesting.  One we have a direct dependency on.  The other we might have indirect (transitive) dependencies on, but dont seem to.

It also looks like System.Memory and System.Buffers got folded into .NET 6 but I dont see that listed in any breaking changes.

**Comment by jzabroski** [comment]
> It also looks like System.Memory and System.Buffers got folded into .NET 6 but I dont see that listed in any breaking changes.

Correction (in post above): System.Memory and System.Buffers have been in-box since .NET Core 2.1 as part of Span being introduced.

**Comment by jacodv** [comment]
Good morning, South Africa (GMT+2) , I think we are worlds apart.   I have a couple of admin tasks for the morning, then I am on this.  

Thanks for the details so far

**Comment by jacodv** [comment]
@jzabroski  this branch compiles and my simple test passes with most of `RazorLight.Tests`.   This is not the solution, I wanted to get a better understanding of the problem, and thus made everything `DotNet 6`.

It does show that there are major compatibility issues going to `DotNet 6`, especially around `ASP.Net Core` and hosting.

I need to get back to my own work.  But will take it up again tomorrow.

**Comment by jacodv** [comment]
@jzabroski , I can confirm that if I start using the converted engine in our code I get "locking" issues.  I think it is in the _cache.Set extension.  I get the same in the RazorLight.Tests project when run on the branch: https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Get-Dotnet6-Working

This is going to be a tough one

**Comment by jzabroski** [comment]
OK, but do you see my point that I suspect major problems may be coming from any of these dependencies:

* Microsoft.Extensions.DependencyModel
* System.Security.Cryptography.Pkcs
* System.Text.Encodings.Web (this one is a guess - I dont see where we use it, though)
* System.IO.FileSystem.AccessContro l(this one is a guess, I dont see where we use it, though)

The first two are the likely culprits.  See: https://github.com/dotnet/runtime/pull/35606 where they refactored how dependencies are tracked from "depproj" projects To be honest, I dont know what a depproj is, but I think its removal is likely related to the above notes about .NET 6 not supporting older libraries and that "harvesting" is no longer due in .NET 6.  See for example https://github.com/dotnet/runtime/issues/29155 where they noted that they should automate harvesting for .NET 5 , then for .NET 6 they decided to remove it altogether because they realized it was a bad idea. But in 35606 they messed up part of the "fix" for removing harvesting.

**Comment by jacodv** [comment]
Possible yes, If you try and support all the frameworks.

But even before we get to the dependencies, just converting to .Net 6 is an issue.  I tried to eliminate all the noise by just focussing on .Net 6 and even that seems to be a challenge.

`await _cache.Set(item.NormalizedKey, taskSource.Task, cacheEntryOptions);` Is one of the culprits

**Comment by jzabroski** [comment]
I'm a little surprised the built-in MemoryCache is causing locking problems.  Can you be more specific?

**Comment by jacodv** [comment]
@jzabroski, I have two relatively full days today and tomorrow but will try and get more detail.  I will focus on this test, it shows the lock.  I am 99% sure it has todo with an `async/await` that included a `.Wait()` or similar.  My branch:
https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Get-Dotnet6-Working shows the behaviour. 
```
[Fact]
public async Task Ensure_Option_Disable_Encoding_Renders_Models_Raw() 
```

**Comment by jzabroski** [comment]
I would recommend the following approach to debugging.

1. See the FAQ where it covers this topic: https://github.com/toddams/RazorLight#im-getting-cannot-find-compilation-library-when-i-deploy-this-library-on-another-server
2. If the error works fine on .NET Core App 3.1 but doesn't work on .NET 6, and the only material difference is the SDK, then I would take a BinLog of both builds and view them line by line in MSBuildStructuredLogViewer utility until you see something that pops out at you.

One thing that pops out at me is this footnote: https://docs.microsoft.com/en-us/aspnet/core/razor-pages/sdk?view=aspnetcore-6.0#runtime-compilation-of-razor-views

> # Runtime compilation of Razor views
> * By default, the Razor SDK doesn't publish reference assemblies that are required to perform runtime compilation. This results in compilation failures when the application model relies on runtime compilation—for example, the app uses embedded views or changes views after the app is published. Set `CopyRefAssembliesToPublishDirectory` to `true` to continue publishing reference assemblies. Both code generation and compilation are supported by a single call to the compiler. A single assembly is produced that contains the app types and the generated views.
> 
> * For a web app, ensure your app is targeting the `Microsoft.NET.Sdk.Web` SDK.

I'm not sure why the RazorLight FAQ recommends setting a different setting to false.  I can see from 2019 there was some advice in this regard, here: https://github.com/dotnet/core/issues/1725#issuecomment-538100548

**Comment by jacodv** [comment]
Thanks @jzabroski, I will try and make time this week to start with the deeper investigation.  I also want to see if I cannot make another project that is even lighter with less moving parts for our needs.

**Comment by jacodv** [comment]
@jzabroski - I think I have the problem.  And it might be on my side.  I use Resharper.  I accidently ran my tests in VS test runner and now I don't see the ERROR.  I have some other issues, but not missing `System.Security.Cryptography.Pkcs` related.

I am not sure if I want it to be Resharper or NOT.

**Comment by jzabroski** [comment]
I hate to be negative, but ReSharper is miles,better than VS Test Runner. It's funny/ironic how bad Microsoft testing is.

Since we have covered so much ground, can you summarize which problems are resolved and which are not?

**Comment by jacodv** [comment]
I agree, and have logged a bug with Jetbrains team.  Not sure if that is a blackhole or not...

With regards to status.  I will start from scratch.  I have basically determined the required / maybe some not new .Net 6 references and versions.  But I think starting over is going to be best and then see what exactly is required to make a .Net 6 project to work.

**Comment by jacodv** [comment]
@jzabroski 

*Short Answer*: I have successfully upgraded to .Net 6 including the latest language version

https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Dotnet6-Upgrade-RazorLight-Tests Show the code and tests

## Running commentary

Please have a look at this [readme](https://github.com/jacodv/RazorLightDotNet6/blob/branch_jdevil-Dotnet6-Upgrade-RazorLight-Tests/DotNet6Status/README.md)

https://github.com/jacodv/RazorLightDotNet6/blob/branch_jdevil-Dotnet6-Upgrade-RazorLight-Tests/DotNet6Status/README.md

I have detailed every step to reproduce as well as what must be changed to enable .Net 6

**Comment by jacodv** [comment]
@jzabroski : What are the next steps to get this out and into Nuget?  I cannot create a pull request due to the 4.7.2 dependencies that does not want to load on VS2022 on my computer.

**Comment by jacodv** [comment]
@jzabroski : This is a roller coaster ride!!!  Next hurdle and this is not **ReSharper** related.  I think I might need to help to understand what is required.  I made sure that ALL the necessary projects have the `PreserveCompilationContext` set to `true`.   The method `internal IReadOnlyList<MetadataReference> Resolve(Assembly assembly, DependencyContext dependencyContext)` receives the test hosts (VSTests or ReSharper) as the assembly and a populated dependency context.  But no paths are resolved.


```
Jdv.Razor.RazorParsingException : One or more errors occurred. (Can't load metadata reference from the entry assembly. Make sure PreserveCompilationContext is set to true in *.csproj file)
```

I have updated the branch with the new *Web Test Project* as well as the README with the detail:

## README:
https://github.com/jacodv/RazorLightDotNet6/blob/branch_jdevil-Dotnet6-Upgrade-RazorLight-Tests/DotNet6Status/README.md

##Branch:
https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Dotnet6-Upgrade-RazorLight-Tests

**Comment by jzabroski** [comment]
Hey man.  

> The method `internal IReadOnlyList<MetadataReference> Resolve(Assembly assembly, DependencyContext dependencyContext)` receives the test hosts (VSTests or ReSharper) as the assembly and a populated dependency context. But no paths are resolved.

Can you explain what you mean here?  Maybe just a screenshot of your watch window with those two variables pinned and expanded?

**Comment by jacodv** [comment]
@jzabroski  Sure, see below:

![image](https://user-images.githubusercontent.com/5814127/144625593-c7ff1c2b-8143-4015-ba86-cdfd17369811.png)

![image](https://user-images.githubusercontent.com/5814127/144625646-91e01797-1cfb-45e2-be12-32c8a297af6e.png)

![image](https://user-images.githubusercontent.com/5814127/144625758-abba20fa-d6e9-454a-b8ac-f62f3b5775ff.png)

**Comment by jzabroski** [comment]
I have a very full day today, but I am going to try hard to get to this tomorrow (weekend time).  Thanks for so much progress.

**Comment by jzabroski** [comment]
@jacodv What version of Resharper are you on so I can ensure we're matching?  I dont think there are any hotfixes to VS2022 yet.

**Comment by jacodv** [comment]
@jzabroski I am on the latest EAP. Also, this is not ReSharper related. I get the same on VS Tests. I will create a new dotnet 6 api this morning

**Comment by jacodv** [comment]
@jzabroski did you manage to have a look over the weekend?

**Comment by jzabroski** [comment]
I did not. My wife had a long TODO list for me :(

**Comment by jzabroski** [comment]
@jacodv I noticed the error message is different than the csproj value.  It says PreserveCompilationContext, whereas your config is PreserveCompilationReferences

**Comment by jzabroski** [comment]
Sorry - I re-read the whole thread and saw your comment from 3 days ago where you set all projects to PreserveCompilationContext.  That sucks.

**Comment by jacodv** [comment]
@jzabroski , Your observation was spot on.  **We have a .Net ready version**.  I am not sure where I got `PreserveCompilationReferences` from, but as soon as I changed it to the correct `PreserveCompilationContext` it worked.  Meaning this branch has a working .Net6 Upgrade (excluding full framework support :-(.

**Would you be able to create new Nuget package for us please?**

*Lastly, We are back to the ReSharper issue though, however I believe this is not RazorLight specific.  I have a ticket open with JetBrains https://resharper-support.jetbrains.com/hc/en-us/requests/3663073.

VS Test Results:
![image](https://user-images.githubusercontent.com/5814127/144983633-9fef259f-e8dc-42a1-895f-f298f7e3fdc8.png)

ReSharper Test Results:
![image](https://user-images.githubusercontent.com/5814127/144983855-8f91b0b8-7cfa-4eb3-926f-50d8d28f9f63.png)

**Comment by jzabroski** [comment]
That's great.  Can I use https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Dotnet6-WebApi6 as instructions for how to update and re-package RazorLight?  Would you like to submit a PR?

I want to set-up a build matrix on GitHub for RazorLight as well.  I hate the fact I've been packaging this on my machine since taking it over from Ivan.

**Comment by jzabroski** [comment]
It looks like since the last time I looked at GitHub Actions, someone has created a GitHub Action for publishing safely to nuget.org - and Luke Lowrey has created a tutorial to set it up.  Should be straight forward to update [.github/workflows/dotnet.yml](https://github.com/toddams/RazorLight/blob/master/.github/workflows/dotnet.yml) using this tutorial: https://lukelowrey.com/use-github-actions-to-publish-nuget-packages/

**Comment by jzabroski** [comment]
@jacodv I'm getting close to automating RazorLight deploys to Nuget.  I was stumped for a bit, but I think the problem was that the root directory is different when you use a build matrix.  Build matrix doesn't even really make sense for pushing a nuget package, so I split the workflows into two separate workflows: [.github/workflows/dotnet-pack-and-push.yml](https://github.com/toddams/RazorLight/blob/master/.github/workflows/dotnet-pack-and-push.yml)

**Comment by jacodv** [comment]
@jzabroski Great news. 

Yes, https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Dotnet6-WebApi6, has all the latest code and successful VS Tests.

Let me know if I can help with anything.  Also, do you think it possible to create an even lighter package if we only use `CompileRenderStringAsync` in our projects?

**Comment by jzabroski** [comment]
I have wanted a lighter package for awhile.  I think the .NET 4.8 support should go in its own csproj as a netstandard2.0 package called RazorLight.Net48.  That should help avoid one of the problems we discussed above, which is when people transitively reference RazorLight from a library that itself targets netstandard2.0 but is meant to target .NET 6 framework.

Can you come up with a general proposal of how you think it should look?  I am not clear on what it would look like with just CompileRenderStringAsync.  In 2020 I did some hotfixes around NullRazorProjectItem to handle some code that was copy-pasted across the Internet, which was the root cause of a LOT of bug reports... so I am a little leery of encouraging people to only use that piece.  I can't remember if the NullRazorProjectItem fix and other fixes were intended to allow for it to work as a stand-alone function.  I had pretty severe COVID in March 2021 and some of my memory isn't quite there, but it's coming back.

**Comment by jzabroski** [comment]
The deploy of rc.4 succeeded but I think the snupkg files may not have made it up.  See: https://github.com/toddams/RazorLight/runs/4452017565?check_suite_focus=true

The diagnostics nuget provides suck.  They started improving them in .NET 5 with better error messages and a troubleshooting page, but they are still not great in the Ci/CD scenario where you can't use Fiddler to spy on http packets.

**Comment by jzabroski** [comment]
Hmm... I am honestly not sure why the GitHub action failed.  It looks like everything succeeded on nuget.org.  I was able to download the package and the symbols.

**Comment by jacodv** [comment]
I saw this  https://github.com/NuGet/Home/issues/8543

**Comment by jacodv** [comment]
@jzabroski  I hope you recover soon.  The new normal, after COVID has some advantages like working form home.  But so many lives were lost and so many people were badly effected, physically and economically.  

W.r.t a use case:
We create a batch of say 10k emails.  All with a razor email body that is captured and stored in a database.  Then when we get to the send stage of our process, we get a JSON document containing the recipient data for each recipient.  We then create messages in a queue with the JSON as well as the email body razor.

The queue messages are processed using a pub/sub mechanism across many servers.  

### Implementation
* Create Singleton for the RazorEngine on each processing server
* Receive message to process
  * Create hash/md5 of the email razor template
  * Check if template has been compiled
  * If not
    * Compile template
    * Store the compiled template in a dictionary with the hash/md as key 
    * Add a timed cache element that will expire if not used for more than 1min
  * Use the compiled template to generate the output given the message body
* Add the transformed email body to the mail object
* Send the email

**Comment by jzabroski** [comment]
Got it.  So you still need RazorEngine and see that whole class as staying put.  That should be fine, as that dependency is where NoRazorProjectItem goes, and handles the scenario where people do crazy things with DI without reading any documentation that annoy me and cause a lot of false positive bug reports.

**Comment by jzabroski** [comment]
Maybe call it RazorLight.NoProjectSystem or similar.  Need to review the code to see exactly what the right name would be.

**Comment by jacodv** [comment]
Maybe `RazorLight.NoProjectEngine`?

**Comment by jzabroski** [comment]
I like that.

**Comment by jacodv** [comment]
@jzabroski , another good thing that came from the exercise is that we helped detect something that would have reached JetBrains' next release.

https://youtrack.jetbrains.com/issue/RSRP-487181

![image](https://user-images.githubusercontent.com/5814127/145521971-c9c90dd5-5d31-4ef0-9d95-35badc2443c9.png)

**Comment by jzabroski** [comment]
@jacodv Impressive work.  You have a great eye for detail and following up.

**Comment by leighmetzroth** [comment]
Hey, just curious how the fix for this is going?

I have also run into this when trying to migrate a project to .net 6. If it would help, I would be happy to branch and do up the PR for it.

**Comment by jzabroski** [comment]
@jacodv Has a list of branches here where he worked through this stuff: https://github.com/jacodv/RazorLightDotNet6/branches

I personally dont have much time to devote to this project any more.  Happy to push through any fixes, though.

**Comment by leighmetzroth** [comment]
@jzabroski Thanks for the quick reply.

I have raised a PR based on the changes from the latest code by @jacodv in https://github.com/jacodv/RazorLightDotNet6/tree/branch_jdevil-Dotnet6-WebApi6

I have excluded some of the extraneous code though.

**Comment by leighmetzroth** [comment]
@jzabroski I noticed you pushed up the update to the nuget version, but won't the new version be `2.0.0-rc.5`?
https://github.com/toddams/RazorLight/commit/7585ff2ef60460dd3968f2b7e15ace16d763216c

**Comment by jzabroski** [comment]
Yes, I was fixing the readme to reflect the current latest.  I forgot to update it.  I might reach out to SImon Cropp to see if he has a GitHub Action for that as part of pushing new versions.

I plan to push to nuget soon enough.  I wanted to see if it was possible to have a RazorLight.net48.csproj for old customers and then we can drop the -rc.  If you want to tackle that yourself, that would expedite things.

**Comment by leighmetzroth** [comment]
@jzabroski I have done a little investigating this morning and I'm a little confused because it seems that it already supports .net Framework 4.8 as it already targets .net Standard 2.0

What I did:
I added a new console app `RazorLight.Sandbox.net48` project and then made it reference the `RazorLight` project. I then included the below code as the main function. Running this it successfully outputs "This is a test Bar" to the console.

Have I missed something about what you are needing?

`
        public static async Task Main(string[] args)
        {
	        var engine = new RazorLightEngineBuilder()
		        .UseMemoryCachingProvider()
		        .Build();

	        var renderedString = await engine.CompileRenderStringAsync(
		        "templateKey",
		        "This is a test @Model.Foo",
		        new
		        {
			        Foo = "Bar",
		        });

			Console.WriteLine(renderedString);
			Console.ReadKey();
        }
`

**Comment by jzabroski** [comment]
The problem I ran into in the past was shipping a netstandard2.0 library rather than a net48 library that references a netstandard2.0 implementation. 

RazorLight.Net48.csproj would support the people stuck on .NET Framework 4.8. It is effectively a copy paste of the RazorLight.csproj but using netstandard2.0 libraries. So the package name is a moniker to guide people to the right package to use for legacy Framework

**Comment by jzabroski** [comment]
I see. I guess my idea only goes so far. Adding a sample is probably best. I guess I should just go ahead and publish.

**Comment by jzabroski** [comment]
@leighmetzroth I remember my thought process now.

1. We are targeting the following TFMs: https://github.com/toddams/RazorLight/blob/master/src/RazorLight/RazorLight.csproj#L3
   a. `netstandard2.0;netcoreapp3.1;net5.0;net6.0`
2. Of those TFMs, `netstandard2.0` isn't a real target, but a pseudo-target.  Because it's a pseudo-target, it can refer to `netcoreapp3.1`, `net5.0`, `net6.0` or `net48`.
   a. `netstandard2.1` started with `.netcoreapp3.0`. Therefore, in some sense, while `netstandard2.0` can refer to `.netcoreapp3.0` and greater, the `PackageReference` elements target AspNetCore 2.1 packages, which are out-of-date and not security patched. Fundamentally, `net48` consumers can't realistically use `netstandard2.1` frameworks (read: `netcoreapp3.0` or later), because there are new features in the runtime like Span.
  b. Therefore, if you're using `netstandard2.0`, you are implicitly saying you plan to use it from a `net48` project entrypoint - but some people want to use `netstandard2.0` to simply mean a common library that can be used in various TFMs across their organization.  This is what gets people into trouble and where they end up opening support requests on RazorLight Issues on GitHub (here), because they don't understand why they are getting runtime assembly conflict issues.

*There*.  Sorry I didn't get all this out on the weekend.  Sometimes I turn my brain down a bit on the weekend.

**Comment by leighmetzroth** [comment]
Given this is blocking a ticket I'm working on that includes an upgrade to .net 6, are we able to publish the new pre-release nuget with .net 6 support so I can get some momentum back up on that work? As you may understand, it's a significant piece of work and each change that's made by my teammates or other tickets I'm working on needs to be merged into this branch; which in itself is costing me a lot of time.

Once my blocked ticket is progressing, I will be able to commit some time to making the change you're suggesting.

Also, is there a way that we could potentially talk through this and the requirements before I tackle it? Going by your profile, you're based in Boston when means that we could jump on a Google Meet early my time, which should be early evening your time.

**Comment by jacodv** [comment]
If I can assist, then I am willing to help.  I will create a VM with all the required frameworks and help.

I am in South Africa and therefor in the middel, let me know if I can be of assistance.

**Comment by jzabroski** [comment]
> Given this is blocking a ticket I'm working on that includes an upgrade to .net 6, are we able to publish the new pre-release nuget with .net 6 support so I can get some momentum back up on that work?

of course. https://github.com/toddams/RazorLight/actions/runs/1848527620  heading away from my desk for next 2 hours but i will follow up to see that it made it up to nuget.org  I picked rc.5 for now - once you confirm it works fine, i will drop the rc tag

**Comment by leighmetzroth** [comment]
Looks like you've been fighting with failures 😔 and the latest failure is because it couldn't find the project.

I've never really dealt with github actions before, but looking at the code [here](https://github.com/alirezanet/publish-nuget/blob/master/index.js) for `alirezanet/publish-nuget`, I can see that it just references `this.projectFile = process.env.INPUT_PROJECT_FILE_PATH` whereas the others reference both the environment variable with and without the `INPUT_` part. I also spotted [this issue](https://github.com/actions/runner/issues/665) where there was a bug where the `INPUT_` was being ignored.

Perhaps we need to switch `PROJECT_FILE_PATH: src\RazorLight\RazorLight.csproj` for `INPUT_PROJECT_FILE_PATH: src\RazorLight\RazorLight.csproj` temporarily?

**Comment by jzabroski** [comment]
It looks like it still failed.  What's odd is I thought I fixed this 2 months ago.  I do remember commenting on this PR https://github.com/brandedoutcast/publish-nuget/pull/62 but I dont remember how I pushed rc.4 - e.g., I think its possible I uploaded everything successfully but the GItHub action was reporting failure

**Comment by jzabroski** [comment]
I think the remaining issue is that the way the code is retrieving the environment var is based on the current process's env variable, which, at the time the step executes, probably is a stale copy of the env variable.  I'm not an expert in GitHub actions, so I am going to try that - mostly explaining what I am doing here to leave breadcrumbs in the future.

**Comment by jzabroski** [comment]
So, there were several issues.

The latest seems to be that pack is failing for some reason... but the GitHub Action isnt checking that commands results for errors (yay!).  Ideally these would be smaller actions, but what do I know.

**Comment by leighmetzroth** [comment]
It was looking so good! Looks like it tried to push as rc-4 again

**Comment by leighmetzroth** [comment]
Does it need the version updated in the `Directory.Build.props`?

**Comment by jzabroski** [comment]
https://github.com/toddams/RazorLight/actions/runs/1855287487

it succeeded but did it publish rc.5? I can't check right now. Driving

**Comment by jzabroski** [comment]
OK, I see the problem.  I forgot I put the version in the src/Directory.Build.props not the top level build props. Should be fixed. It will be published as rc.6 unfortuantely

**Comment by jzabroski** [comment]
@leighmetzroth @jacodv We did it!  https://www.nuget.org/packages/RazorLight/2.0.0-rc.6

**Comment by jacodv** [comment]
Well done @jzabroski

**Comment by leighmetzroth** [comment]
It is strange that it's not showing there, but it does show in the Frameworks section below. Maybe there's some process that updates it later???

Anyway, the good news is that the new version works in a test .NET 6 project. The bad news is that it looks like I'm about to be in "DLL Hell" as it was called earlier in this thread as it appears that something is pulling in the old version of `System.Security.Cryptography.Pkcs` in our solution. I'm fairly sure I know what it is (an internal NuGet that we have that's still referencing .NET 2.2).

![image](https://user-images.githubusercontent.com/34053918/154595593-608e35f9-78a3-416c-9c71-12bdd47fbf4e.png)

**Comment by jacodv** [comment]
Fantastic!

I am resuming our upgrade next week!

**Comment by leighmetzroth** [comment]
<img src="https://media0.giphy.com/media/CuMiNoTRz2bYc/giphy.gif"/>

After a heap of digging, I found that we have a transitive dependency on `System.Security.Cryptography.Pkcs` via MailKit (it has a dependency on MimeKit, which takes the dependency on `System.Security.Cryptography.Pkcs`).

After upgrading to the latest MailKit, I still had issues and ended up trawling through all of our own internal nugets (of which quite a few were in dire need of upgrading to the latest frameworks too 😬 ). After all of that, it still didn't work; uninstalling MailKit and commenting out the code that sends the emails using it, it did work....

I did some more Google-foo and eventually landed on [this](https://stackoverflow.com/questions/70236519/invalidoperationexception-cannot-find-compilation-library-location-for-package), which has an answer for adding `<GenerateRuntimeConfigDevFile>true</GenerateRuntimeConfigDevFile>` to the csproj files (the same stack of projects that need the `PreserveCompilationContext` set to true). I tried that, and it worked. So I have a feeling it's got something to do with the refactor they made to these DLLs and with the removal of the runtime config dev file, it has left projects like this unable to discover these dependencies for some reason.

Is it worth adding some instructions to the FAQ for this? If you think it is, just let me know and I'll write up the instructions and PR it in.

**Comment by jzabroski** [comment]
Yes, but... Please mention the actual issue rather than StackOverflow: https://github.com/dotnet/sdk/issues/16818

**Comment by jacodv** [comment]
WOW, I added that element to the project file weeks ago, but it was to to get ReSharper tests to work.  Who would have thought....

**Comment by jzabroski** [comment]
@jacodv It's funny, isn't it.  We were told that getting rid of Assembly Binding Redirects and shadow copy assemblies would make dotnet core infinitely easier than .NET Framework.  And here we are, playing with a bunch of new knobs.  Feels an awful lot like the old new thing, doesn't it?  I do think its slightly improved in some core ways, but everything outside that core shell is now a mess.  This has been my biggest takeaway working with .NET Core and later - dynamically loading plugins without a well-defined assembly load context is a bad idea.  But the problem is the compiler framework doesn't have any way to hint at what dynamic load targets you might plan, e.g. if you target netstandard2.0.

@leighmetzroth Let me know if you still plan to do the P.R.

**Comment by leighmetzroth** [comment]
@jzabroski I am planning on doing it, just a bit busy right now between sick kids at home and having to take time off, and getting this .NET 6 upgrade completed and tested so we can roll it out to our customers.

**Comment by jzabroski** [comment]
Sounds good.

**Comment by jacodv** [comment]
@leighmetzroth @jzabroski - We have successfully upgraded to .Net 6.  Thanks for all the help.

**Comment by jacodv** [comment]
Nope, Added the required project attributes [`PreserveCompilationContext`, `GenerateRuntimeConfigDevFile`] and all good.

**Comment by mikart143** [comment]
So for people still having the issue the ultimate solution for this is setting PreserveCompilationContext to true and adding ExcludingAssembly like:
```
            services.AddRazorLight(() =>
            {
                return new RazorLightEngineBuilder()
                    .UseFileSystemProject(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location))
                    .UseMemoryCachingProvider()
                    .ExcludeAssemblies(typeof(System.Security.Cryptography.Pkcs.AlgorithmIdentifier).Assembly.GetName().Name) // Workaround for loading CompileLibraries
                    .EnableDebugMode(Debugger.IsAttached)
                    .Build();
            });
```

If you will try to Exclude Assembly with FullName It will not work. Additonally the RazorLightDependancyEngineBuilder is broken and all ExcludedAssemblies are ignored further in DefaultMetadataReferenceManager. @jzabroski

**Comment by jzabroski** [comment]
> Additonally the RazorLightDependancyEngineBuilder is broken and all ExcludedAssemblies are ignored further in DefaultMetadataReferenceManager. @jzabroski

Can you explain what you mean by opening a new issue?
